docs/project_proposal.md

```diff
@@ -0,0 +1,116 @@
+# Starlark Language Server Proposal
+
+Do you wish that your editor/IDE provided fancy capabilities for [Starlark] (e.g., auto complete, go
+to definition, hover information)? If so, read on and help us get to our funding goal to make this
+project a reality.
+
+| Quick Facts  |                                                                                                                                                          |
+| ------------ | -------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| What         | Implement a language server for [Starlark].                                                                                                              |
+| Why          | Fancy features like go to definition, hover information and auto complete in your favorite editor/IDE. (see [Language Server Protocol] for more details) |
+| Who          | [Chuck Grindel]                                                                                                                                          |
+| When         | As soon as we meet our funding goals.                                                                                                                    |
+| Language     | Implement the server using [Go].                                                                                                                         |
```

Comment:
I think there are 3 reasonable choices for implementation:

- Java, as part of Bazel JVM daemon

- Go, as proposed here

- Rust

It would be interesting to have a discussion regarding pros/cons of each approach. I am not convinced that `Go` is the best solution here given the `Java` approach is more tightly integrated with Bazel and `Rust` offers better performance and a more advanced parser (supporting type hint, which gona be needed to provide Code Completion down the line).

Reply:
Perhaps I should update the proposal to include my reasoning for using Go. My plan is to leverage [Starlark Go](https://github.com/google/starlark-go) and [LSP in Go](https://github.com/go-language-server/protocol). Plus, it could be easier to leverage code from [buildifier](https://github.com/bazelbuild/buildtools/tree/master/buildifier) and [bulldozer](https://github.com/bazelbuild/buildtools/tree/master/buildozer), given that they are implemented in Go.

About using Java, the only real pro would be to leverage existing Bazel code. However, I am leary of trying to hook into Bazel internals. It feels brittle.

About using Rust, TBH, I did not consider it. My focus has primarily been on Go to leverage existing libraries as I stated above. Not having done any implementation in Rust, I don't have a good feel for the benefits.

Reply:
Rust starlark implementation (by Meta) already included an LSP implementation that is mostly functional https://github.com/facebookexperimental/starlark-rust/tree/main/starlark/src/lsp. This is different from Buck2's own starlark LSP, which is extended this LSP to provide Buck2's native rules and builtin functions. So you could already use this LSP implementation today to navigate within 1 file in a Bazel project.

The LSP is also built on top of the same library that made rust-analyzer, the canonical LSP for rust ecosystem and battle-tested by Meta.

So I suspect you would get a much higher ROI by using Rust over Go in this regard.

> About using Java, the only real pro would be to leverage existing Bazel code. However, I am leary of trying to hook into Bazel internals. It feels brittle.

I would say the same for buildifier and Starlark Go code base having worked with them in the past.
However, there is no denying that the source of truth _IS_ Bazel's Java.

---

Please feel free to pick which ever is more comfortable to you. I think they are all good approaches with different tradeoffs.

Reply:
@sluongng Ah. I was not aware of the Rust Starlark LSP. I will check it out.